### ISSUE TITLE
AI Function Node Iteration

### ISSUE THREAD

**Issue by joepavitt** [FlowFuse/flowfuse#4073]
### Description

- [x] Build a REST API service that wraps the same intelligence running in the `function-gpt` node.
- [x] ~~Build a new `function-ai??` node that will run on all NR instances on FlowFuse Cloud (and maybe self-hosted if we can work out smooth way of deploying)~~
- [x] Plugin for Node-RED that tightly integrates with the editor
   - [x] https://github.com/FlowFuse/admin/issues/320
   - [x] https://github.com/FlowFuse/nr-assistant
- [x] Support instances
   - [x] FlowFuse/nr-launcher#255
   - [x] FlowFuse/flowfuse#4083
- [ ] Automation, integration, configuration
   - [x] https://github.com/FlowFuse/helm/issues/425 
   - [x] https://github.com/FlowFuse/nr-assistant/issues/6
   - [x] https://github.com/FlowFuse/nr-launcher/issues/259
   - [x] FlowFuse/CloudProject#457
       - [x] Staging
       - [x] Production
- [ ] Support devices
   - [ ] Add necessary config settings to device-agent for nr-assistant
   - [ ] Update FF snapshot to include nr-assistant package in snapshot
- [ ] Follow up tasks
   - [x] FlowFuse/flowfuse#4101
   - [ ] FlowFuse/website#2286

### Which customers would this be available to

All

### Have you provided an initial effort estimate for this issue?

I have provided an initial effort estimate

**Comment by Steve-Mcl** [comment]
### Design

Having been working on the 1st part (API) and I have stuck with Open AI.  

Using `functions` (example: https://www.datacamp.com/tutorial/open-ai-function-calling-tutorial) I have got it to return parsable data including any additional libs

example response:
![image](https://github.com/FlowFuse/flowfuse/assets/44235289/1dd9999c-c49f-4281-9984-b6b330b42834)

This lead me to the thought - what if we didnt have a 3rd full duplicate of the core function node but instead, made this available to FF based Node-RED instances via a plug in?

The plugin, when called could return all of the elements required to simply generate a new function node for the canvas.

Additionally, when the function node is already created in the workspace, we can still call the API via the plugin by adding a prompt in the monaco editor.

Thoughts?

**Comment by joepavitt** [comment]
So, from a UX perspective, two entry points:

- Plugin sidebar? Ask any request, get a new function node
- Core function node has extra method/input in it as per our node and they can interface with that?

It feels more intuitive and obvious to a user if it's an explicit text input available to use, rather than them having to write comment-based prompts for example

**Comment by joepavitt** [comment]
Agree with the plugin direction, rather than _another_ node. Coupling tightly with existing function node would be preferable

**Comment by ZJvandeWeg** [comment]
@joepavitt Which tier of customer is this available to?

**Comment by joepavitt** [comment]
Would be all - not sure why that was missed

**Comment by Steve-Mcl** [comment]
Just sharing some progress and a request for feedback

### Current operational state (rough n ready)

#### Generating a function node
![chrome_txet05w8ty](https://github.com/FlowFuse/flowfuse/assets/44235289/9725b689-4a12-41f3-bcc0-e9bd017828ad)


#### Generating a flow
![chrome_mJLVn7v7Xx](https://github.com/FlowFuse/flowfuse/assets/44235289/319b846c-ecac-45a1-84f5-67ea7d174121)

#### Generating JavaScript while inside a function 
![chrome_VEWqsMUe4C](https://github.com/FlowFuse/flowfuse/assets/44235289/55d8f74a-05d4-4027-9701-0032de4e8ba7)


#### Notes
You might wonder why there is a separate "Flow Builder" and "Function Node Builder"...
In short, asking the AI to make a function node is super reliable. I provide system prompts to generate plain JavaScript (suitable for a node-red function) and use the OpenAI `function + schema` to respond with a JS object containing number of outputs, function code, any npm modules it may have used etc. Then, in FF I hydrate that into a well formed Node-RED JSON array containing a single function node.  This works very well.

The flow builder is more more suseptible to "halucinations" and gets thigs wrong _very often_

Lastly the editor "Ask FlowFuse 🪄" uses something called a Code Lens (monaco feature) and calls to the same endpoint as the Function Node builder. This means we can do similar things like adjust the output count and add libs to the setup tab.  


<br>

### Accessing these menu items (please vote which option)

I could not decide (discussed with @joepavitt) how to surface these. we agreed i would post screenshots for comparison. 
Please take a look and chose what feels best.

#### Option 1 - front and center (not easy to miss)
![image](https://github.com/FlowFuse/flowfuse/assets/44235289/b51c81e9-cb66-45c9-8e74-47c35b7578ea)

#### Option 2 - top level side menu
![image](https://github.com/FlowFuse/flowfuse/assets/44235289/4d68f370-6e00-4614-a9ed-cb483740becd)

#### Option 3 - added to an existing side sub menu (I chose "view", but it could be any)
![image](https://github.com/FlowFuse/flowfuse/assets/44235289/c66f1f60-401d-4bbd-9aa3-35d6c3f0af86)

### Other feedback would be great too

The menu items, dialogs titles / descriptions of the feature are very much thown together. Some feedback and suggestions are welcome.

* Global operations (flow builder / function Node builder)
  * Menu item titles
  * Dialog titles
  * Prompt descriptions on dialog body
  * AI Magic Icon in top header (linked issue https://github.com/FlowFuse/website/issues/2258) 
* Inline code action
  *  "Ask flowfuse 🪄`" Good?  Bad? Other?
  * Dialog title
  * Prompt description on dialog body

**Comment by Steve-Mcl** [comment]
Future iteration though dump

Via Monaco Editor...
1. Code lens could be used on JSON editor  - "Generate an array of 10 objects containing random names and dates"
2. Code lens could be used on the template editor for the various code styles (JSON, CSV, JavaScript, SQL)
3. Code Lens could be used in the Dashboard 2.0 ui-template for assistance building templates (untested at this time)

Additional menu items:
1. Select a bunch of nodes and "Ask FlowFuse what this flow does"
2. More specific prompts that can have fine tuned system hints like "Add an endpoint", "Make a HTTP Request"
   3. Under the hood, this would work similar to the function node builder where specific system hints asked for specific structured response then we hydrate the response into a well formed flow

**Comment by knolleary** [comment]
Given the hallucinations, lets keep the scope on the Function node for this iteration (so we can get it released this week) and do the flow builder for v2.

Of the three options, I think Option 1 is probably the best to begin with to get this feature noticed. Burying it under the menu will make it hard to discover.

The Code Lens piece looks really nice 👍🏻 

For naming, I'd suggest `@flowfuse/nr-assistant` - I could see this expanding out to other features which don't necessarily have to be anything to do with AI, but are there to help the user.

Where can I see details of the rest api this uses? We need to make sure the security around it is appropriate.

I presume this will work on the Device Agent as well?

**Comment by Steve-Mcl** [comment]
@knolleary I had written most of the below when your comment landed (so it may seem a bit mixed up.


> Where can I see details of the rest api this uses? We need to make sure the security around it is appropriate.

Regarding integration with FF platform.

currently, this plugin is under local development and uses an endpoint hosted in a Node-RED instance on FFC.  This make iteration and improvement super quick.

However, to implement this securely, currently, the endpoint is protected using a single token setup on the instance security settings.

I cannot immediately think of a way to avoid having to bring this endpoint and API into the FF platform.

Can I grab you for 5 mins to discuss a scalable and suitable direction for integration with FF core please?



> I presume this will work on the Device Agent as well?

It was not mentioned in scope and is a question I think you have just answered ;)
I also think that ties up with the previous point above.
 

> For naming, I'd suggest @flowfuse/nr-assistant

Yes, makes sense & I love it too (bonus).


> Given the hallucinations, lets keep the scope on the Function node for this iteration (so we can get it released this week) and do the flow builder for v2.

I agree completely. will remove that menu item for 1st iteration.

> Of the three options, I think Option 1 is probably the best to begin with to get this feature noticed. Burying it under the menu will make it hard to discover.

Perfect.

**Comment by knolleary** [comment]
Summary of a discussion @Steve-Mcl and I had:

For first iteration:

1. Adding api to core platform to expose the assistant apis. Will be secured based on instance/device token.
2. OpenAI config provided in platform `yml` file to enable this feature

This will then allow us to enable this feature for FFC users.

For self-hosted, in the first iteration, they will be able to provide their own OpenAI tokens in their `yml` file to unlock this feature.

Future iteration - provide a public API that self-hosted instances can access to enable this feature. The api request will include their license identifier which we can use to validate the request is a valid customer.

Note - self-hosted on air-gapped networks will not be able to make use of this feature.

**Comment by Steve-Mcl** [comment]
Regarding: HTTP requests made by the nr-assistant plugin (from runtime -> AI API)

In PoC, I simply used Node built-in fetch which works fine however we have some users stuck on node 14/16 (on devices?)  and fetch is Node 18 min (and even then it is experimental until Node 21)

I noted that nr-tools uses the nifty little `undici` lib - but this is around 1.2mb installed and currently requires node >= `18.17`

There is also a super small (160kb installed) `ky` (by the `got` folk) but it would be yet another/different lib 

So I have decided to stick with got for the plugin for the first iteration but I am struggling to justify NOT going with v11 of `got`

Do we: 
* Stay with v11 + `require` (like device agent, nr-files, and some others) 
* OR 
* Follow Node-RED/nr-project-nodes lead and go with v12.6.0  + dynamic `import`
* OR
* `undici` (like `nr-tools`) meaning users of the plugin must be on Node 18.17 min


My gut says since the plugin runs in the Node-RED runtime we should match up with Node-RED & use `12.6.0` + dynamic import HOWEVER that means the plugin is limited to later versions of node-red?

**Comment by knolleary** [comment]
Just to be accurate, nr-tools uses undici 5.x which supports older Node.js versions.

`got` v11 is tried and tested. It isn't necessary my first choice these days, but lets stick with what works.

**Comment by knolleary** [comment]
@Steve-Mcl we will need an issue on the helm repo to add the templated settings to the flowforge.yml file - this is what the cloudproject issue will then be able to make use of when we publish the release.

**Comment by Steve-Mcl** [comment]
@knolleary raised here: https://github.com/FlowFuse/helm/issues/425

**Comment by joepavitt** [comment]
@Steve-Mcl can you drop you status update here when available please?

**Comment by Steve-Mcl** [comment]
> @Steve-Mcl can you drop you status update here when available please?

* API in FF merged
* basic automations have just been merged in nr-assistant. 
  * That means i can now tag a release - About to do.
  * That unblocks the nr-launcher PR https://github.com/FlowFuse/nr-launcher/pull/256 - Next task
* follow up tasks for release process are being kindly handled by @ppawlowski 
   * https://github.com/FlowFuse/nr-assistant/issues/6  (discussed)
   * https://github.com/FlowFuse/helm/issues/425 (in progress)
   * https://github.com/FlowFuse/CloudProject/issues/457 (thursday item)
   
following that, I will be finishing up device agent integration - but that is non blocking of the release for instances.

**Comment by joepavitt** [comment]
Closing this out as it's now in production

**Comment by Steve-Mcl** [comment]
> Closing this out as it's now in production

Should we reopen for device agent support or generate new issue Joe?

**Comment by knolleary** [comment]
@Steve-Mcl new issue that is specifically scoped to agent support
